cvefeed/diff.go

```diff
@@ -0,0 +1,301 @@
+	chunkDescription chunk = 1 << iota
+	chunkScore
+)
+
+var chunkKind = [chunkMaxShift]ChunkKind{
+	ChunkDescription,
+	ChunkScore,
+}
+
+func chunkShiftByName(kind ChunkKind) int {
+	for i, v := range chunkKind {
+		if v == kind {
+			return i
+		}
+	}
+	return chunkMaxShift
+}
+
+type diffEntry struct {
+	id   string
+	bits chunk
+}
+
+type diffFeed struct {
+	name string
+	dict Dictionary
+}
+
+type diffInput struct {
+	a, b diffFeed
+}
+
+type diff struct {
+	diffInput
+}
+
+func newDiff(i diffInput) *diff {
+	return &diff{
+		diffInput: i,
+	}
+}
+
+// DiffStats is the result of a diff.
+type DiffStats struct {
+	diff *diff // back pointer to the diff these stats are for
+
+	numVulnsA, numVulnsB int
+
+	aNotb     []string // ids of vulns that are in a but not in b
+	bNota     []string // ids of vulns that are in b but not in a
+	entries   []diffEntry
+	bitCounts [chunkMaxShift]int
+}
+
+// NumVulnsA returns the vulnerability in A (the first input to Diff).
+func (s *DiffStats) NumVulnsA() int {
+	return s.numVulnsA
+}
+
+// NumVulnsB returns the vulnerability in A (the first input to Diff).
+func (s *DiffStats) NumVulnsB() int {
+	return s.numVulnsB
+}
+
+// VulnsANotB returns the vulnerabilities that are A (the first input to Diff) but
+// are not in B (the second input to Diff).
+func (s *DiffStats) VulnsANotB() []string {
+	return s.aNotb
+}
+
+// NumVulnsANotB returns the numbers of vulnerabilities that are A (the first input
+// to Diff) but are not in B (the second input to Diff).
+func (s *DiffStats) NumVulnsANotB() int {
+	return len(s.aNotb)
+}
+
+// VulnsBNotA returns the vulnerabilities that are A (the first input to Diff) but
+// are not in B (the second input to Diff).
+func (s *DiffStats) VulnsBNotA() []string {
+	return s.bNota
+}
+
+// NumVulnsBNotA returns the numbers of vulnerabilities that are B (the second input
+// to Diff) but are not in A (the first input to Diff).
+func (s *DiffStats) NumVulnsBNotA() int {
+	return len(s.bNota)
+}
+
+// NumDiffVulns returns the number of vulnerability that are in both A and B but
+// are different (eg. different description, score, ...).
+func (s *DiffStats) NumDiffVulns() int {
+	return len(s.entries)
+}
+
+// NumChunk returns the number of different vulnerabilities that have a specific chunk.
+func (s *DiffStats) NumChunk(chunk ChunkKind) int {
+	return s.bitCounts[chunkShiftByName(chunk)]
+}
+
+// PercentChunk returns the percentage of different vulnerabilities that have a specific chunk.
+func (s *DiffStats) PercentChunk(chunk ChunkKind) float64 {
+	return float64(s.bitCounts[chunkShiftByName(chunk)]) / float64(len(s.entries)) * 100
+}
+
+func diffDetails(s *schema.NVDCVEFeedJSON10DefCVEItem, bit chunk) bag {
+	var v interface{}
+
+	switch bit {
+	case chunkDescription:
+		v = bag{
+			"description": englishDescription(s),
+		}
+	case chunkScore:
+		v = s.Impact
+	}
+
+	var data map[string]interface{}
+	tmp, _ := json.Marshal(v)
+	_ = json.Unmarshal(tmp, &data)
+
+	return data
```

Comment:
there's some inconsistency in usage of `type bag map[string]interface{}` and raw `map[string]interface{}`. I'd ditch the former, personally, although, this type saves a lot of typing along the way, so it's not a strong opinion.

Reply:
used bag here :)